cmd/jujud/agent/machine.go

```diff
@@ -710,62 +707,35 @@ var (
 	newBroker     = broker.New
 )
 
-// startAPIWorkers is called to start workers which rely on the
-// machine agent's API connection (via the apiworkers manifold). It
-// returns a Runner with a number of workers attached to it.
-//
-// The workers started here need to be converted to run under the
-// dependency engine. Once they have all been converted, this method -
-// and the apiworkers manifold - can be removed.
-func (a *MachineAgent) startAPIWorkers(apiConn api.Connection) (_ worker.Worker, outErr error) {
-	// CAAS agents do not have any api workers.
+func (a *MachineAgent) machineStartup(apiConn api.Connection, logger machine.Logger) error {
+	logger.Tracef("machineStartup called")
+	// CAAS agents do not have machines.
 	if a.isCaasAgent {
-		return nil, dependency.ErrUninstall
+		return nil
 	}
-	agentConfig := a.CurrentConfig()
 
 	apiSt, err := apiagent.NewState(apiConn)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
 	}
 	entity, err := apiSt.Entity(a.Tag())
 	if err != nil {
-		return nil, errors.Trace(err)
-	}
-
-	runner := worker.NewRunner(worker.RunnerParams{
-		IsFatal:       agenterrors.ConnectionIsFatal(logger, apiConn),
-		MoreImportant: agenterrors.MoreImportant,
-		RestartDelay:  jworker.RestartDelay,
-		Logger:        logger.Child("runner"),
-	})
-	defer func() {
-		// If startAPIWorkers exits early with an error, stop the
-		// runner so that any already started runners aren't leaked.
-		if outErr != nil {
-			_ = worker.Stop(runner)
-		}
-	}()
-
-	// Perform the operations needed to set up hosting for containers.
-	if err := a.setupContainerSupport(runner, apiConn, agentConfig); err != nil {
-		cause := errors.Cause(err)
-		if params.IsCodeDead(cause) || cause == jworker.ErrTerminateAgent {
-			return nil, jworker.ErrTerminateAgent
-		}
-		return nil, errors.Annotate(err, "setting up container support")
+		return errors.Trace(err)
 	}
-
 	// Report the machine host name and record the agent start time. This
 	// ensures that whenever a machine restarts, the instancepoller gets a
 	// chance to immediately refresh the provider address (inc. shadow IP)
 	// information which can change between reboots.
 	hostname, err := getHostname()
 	if err != nil {
-		return nil, errors.Annotate(err, "getting machine hostname")
+		return errors.Annotate(err, "getting machine hostname")
 	}
 	if err := a.recordAgentStartInformation(apiConn, hostname); err != nil {
-		return nil, errors.Annotate(err, "recording agent start information")
+		return errors.Annotate(err, "recording agent start information")
+	}
+
+	if err := a.setupContainerSupport(apiConn, logger); err != nil {
+		return err
```

Comment:
Should we annotate this?

Reply:
This method can return errors from the dependency pkg. I've noticed that if wrapped, the dependency engine doesn't handle them as expected. I added logging messages to explain why we were calling ErrUninstall. Check that they make sense so we can find answers in the future.